src/daft-connect/src/op/execute/root.rs

```diff
@@ -0,0 +1,72 @@
+use std::{collections::HashMap, future::ready};
+
+use common_daft_config::DaftExecutionConfig;
+use futures::stream;
+use spark_connect::{ExecutePlanResponse, Relation};
+use tokio_util::sync::CancellationToken;
+use tonic::{codegen::tokio_stream::wrappers::UnboundedReceiverStream, Status};
+
+use crate::{
+    op::execute::{ExecuteStream, PlanIds},
+    session::Session,
+    translation,
+};
+
+impl Session {
+    pub async fn handle_root_command(
+        &self,
+        command: Relation,
+        operation_id: String,
+    ) -> Result<ExecuteStream, Status> {
+        use futures::{StreamExt, TryStreamExt};
+
+        let context = PlanIds {
+            session: self.client_side_session_id().to_string(),
+            server_side_session: self.server_side_session_id().to_string(),
+            operation: operation_id,
+        };
+
+        let finished = context.finished();
+
+        let (tx, rx) = tokio::sync::mpsc::unbounded_channel::<eyre::Result<ExecutePlanResponse>>();
+        std::thread::spawn(move || {
+            let result = (|| -> eyre::Result<()> {
+                let plan = translation::to_logical_plan(command)?;
+                let logical_plan = plan.build();
+                let physical_plan = daft_local_plan::translate(&logical_plan)?;
+
+                let cfg = DaftExecutionConfig::default();
+                let results = daft_local_execution::run_local(
```

Comment:
Instead of spawning a new thread to run the execution engine (which internally spawns it's own thread), you can have it return you a receiver and the handle to the thread, which you can consume asynchronously from this function.

Reply:
pretty sure there are other thingsi n here like daft_local_plan::translate that require being in another thread unless you recently changed this.

Reply:
pretty sure there are other thingsi n here like daft_local_plan::translate that require being in another thread unless you recently changed this.

Reply:
resolving this as no response.